### ISSUE TITLE
output of extract_feature is a nan tensor when eval model

### ISSUE THREAD

**Issue by rudvlf0413** [lukemelas/EfficientNet-PyTorch#185]
Hello, I'm implementing an image captioning model using the EfficientNet model (efficientnet-b5).
When I tried to extract features using the EfficientNet model with eval mode, I got a nan tensor.

Following codes generate the model:
```
image_size = EfficientNet.get_image_size(args.efficientnet)
efficientnet = EfficientNet.from_pretrained(args.efficientnet, advprop=True)
feature_dim, feature_size = efficientnet.extract_features(
    torch.FloatTensor(1, 3, image_size, image_size)).shape[1:3]

model = TransformerCaptioning(voca_num, pad_idx=args.pad_idx, bos_idx=args.bos_idx, 
                eos_idx=args.eos_idx, max_len=args.max_len, feature_dim=feature_dim, feature_size=feature_size, d_model=args.d_model, n_head=args.n_head, dim_feedforward=args.dim_feedforward, num_encoder_layer=args.num_encoder_layer, num_decoder_layer=args.num_decoder_layer, dropout=args.dropout, embedding_dropout=args.embedding_dropout)
model = model.to(device)
model.extractor = efficientnet.to(device)
```

When I run codes as

```
model.extractor.eval() # extractor is EfficientNet
imgs, caps = next(iter(train_loader))
imgs = imgs.to(device)

with torch.no_grad():
    test = model.extractor.extract_features(imgs)
    
print(test)
```

then the nan tensor is returned as

<img width="409" alt="image" src="https://user-images.githubusercontent.com/6302455/82783873-5bc4d580-9e9a-11ea-8a20-b373057eb1f1.png">

However, when I changed the eval mode to train mode, the normal tensor was returned as

```
model.extractor.train() # extractor is EfficientNet
imgs, caps = next(iter(train_loader))
imgs = imgs.to(device)

with torch.no_grad():
    test = model.extractor.extract_features(imgs)
    
print(test)
```
<img width="582" alt="image" src="https://user-images.githubusercontent.com/6302455/82783795-27511980-9e9a-11ea-93dd-21f24247aa95.png">

Do you know why the EfficientNet model returns a nan tensor when eval mode?

**Comment by Johann-Huber** [comment]
Hey ! I am facing the exact same issue ; how did you fix it ?

**Comment by rudvlf0413** [comment]
@Johann-Huber I confuse about how I solved this problem after a long time.
However, I might solve this problem by changing the code for the model initialization.

For example, you can change this code
'''
model = TransformerCaptioning(...)
model = model.to(device)
model.extractor = efficientnet.to(device)
'''

as
'''
model = TransformerCaptioning(...)
model.extractor = efficientnet
model = model.to(device)
'''

With some initialization ways, I think the problem occurring NaN values happens.

**Comment by Johann-Huber** [comment]
Hi @rudvlf0413 , thank you for the quick answer !

I've finally figure out where the problem comes from. Actually, it is due to the BatchNorm2d layers, and the way pytorch handles them in train() and eval() mode.

In train() mode, pytorch computes the BN trainable parameters using each batch, as described in the paper. At each iteration, those parameters are updated using the current batch mean and std, as well as the previous batches mean and std (using an exponential moving average).

In eval() mode, there shouldn't be batch mean & std computing, as inference shouldn't require an full batch to do prediction. Therefore, pytorch uses the batch mean & std computed during training.

Problem occurs if the distribution of the dataset used for inference is not close enough to the dataset used for the training iterations. In such case, the BN trainable parameters does not reflect the actual batch distribution, leading to overestimated activations. The feature maps values increase exponentially in the network, eventually returning NaN values.

So in a sense, it is related to initialization (dataset & first training iterations in particular). Unfortunately, I was not able to fix the issue by playing with the model initialization, and the way its loaded into the training device. If you remember something about it, I would be interested to know !

**An easy way to fix the problem is described in [this discussion](https://discuss.pytorch.org/t/model-eval-gives-incorrect-loss-for-model-with-batchnorm-layers/7561/40), and basically consists of setting `track_running_stats=False` in the BatchNorm2d layers.**

As I said above, it is efficient to avoid the validation loss explosion issue, but in a partially satisfying manner. Other solutions proposed in the mentioned discussion did not work for me, I will investigate it a bit further to figure out how this problem is handled in other frameworks.

Thank you again for your help !

**Comment by baopanda** [comment]
> explosion

Hi @Johann-Huber 
Thanks for your reply and answer. It really help me because I also face the same issue but do you found a nice solution for it? Can you share it to me?

Thank you in advance.

**Comment by Johann-Huber** [comment]
@baopanda With hindsight, I would say that the big problem here is the quality of the training data. As said above, the problem occurs if **there is a distribution shift between training and validation sets.** 

I suggest you to first **make sure the problem actually comes from here** : a good way to do so is to start several trainings with different seeds for you train/valid split, and check if the estimated values for mean/std at batchnorm layer are far from each others between runs.

**It would basically mean your data should be augmented** to properly train your model.

In case you are restricted to a specific dataset (and if data augmentation is not enough), then the `track_running_stats=False` trick should do the job. But be aware that it is a dangerous workaround, as you might face the exact same problem when trying to deploy your model on a test dataset.

**Comment by baopanda** [comment]
@Johann-Huber Yeppp Thank you for your quick answer.
